### PR TITLE
Ordering/ordering

### PR DESCRIPTION
**Pull request by razorao**
### How Ordering is implemented:

- **Sequences:**
     We track two keys:
     1. `offset-key`: Tracks the sequence of messages in a given ordering key. Used to assign sequence numbers for messages.
     2. `offset-status-key`: Tracks the status of the messages that have been read from primary topic. If for a given ordering key,  a _FAILURE_ is recorded, the subsequent messages are paused until the status wither becomes _SUCCESS_ or _DLQ_

- **Ordering Pull:**
      1. Pull messages from broker
      2. For a primary topic message with ordering key, assign sequence number and previous sequence number.
      3. For all primary topic messages with ordering key, fetch the recorded status. If any previous message in order has failed, pause the primary consumer. Do not pause the retry topic consumer. 
      4. The messages that can be delivered, is returned.
      5. The messages that are pulled but paused are cached in memory.
      6. Deliverable messages are added in memory for offset management.

- **Ack/Nack**
     1. The status of the messages are recorded in memory against their offsets
     2. Messages will be evicted from memory

- **Eviction from Memory:**
    1. For a given topic, we fetch all the offsets which have received ack/nack starting from min offset
    2. If a message has received an ACK, it is not retried. If it has received NACK, it is retried.
    3. If for a given message of an ordering key, there exists a previous message with _FAILURE_ status, the message is retried even if an ACK is received for the message.
    4. The messages are committed to the topic.
    5. The messages are evicted from memory.

- **Deletion of status keys:**
    1. During eviction, for a given ordering key, check the last read/pulled sequence number.
    2. Fetch the status of the ordering key.
    3. If the status is _SUCCESS_ or _DLQ_ and the offset is same as the last pulled message, then the keys for the ordering key can be deleted.

### CODE REVIEW

File: internal/subscriber/subscriberimpl.go [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+			msgID := peek.MsgID
+			if _, ok := metadata.consumedMessages[msgID]; !ok {
+				// check if message is present in-memory or not
+				continue
+			}
+			msg := metadata.consumedMessages[msgID].(messagebroker.ReceivedMessage)
+
+			// NOTE :  if push to retry queue fails due to any error, we do not delete from the deadline heap
+			// this way the message is eligible to be retried
+			s.retry(ctx, s, s.consumer, s.retrier, msg, errChan)
```
Comment (sunnyagg) [inline internal/subscriber/subscriberimpl.go:263]: shouldn't we call nack here ?

**Review by vnktram — APPROVED** [review]

File: internal/offset/model.go [hunk 2]
```diff
@@ -39,3 +45,17 @@ func (m *Model) Key() string {
 func (m *Model) Prefix() string {
 	return common.GetBasePrefix() + offsetPrefix
 }
+
+// Key returns the key for storing in the registry
+func (m *Status) Key() string {
+	key := m.Prefix() + m.Topic + subscriptionPrefix + m.Subscription + partitionPrefix + strconv.Itoa(int(m.Partition))
+	if len(m.OrderingKey) > 0 {
+		key = key + orderingPrefix + m.OrderingKey
```
[the remark this answers is not in the document]
Comment (razorao) [reply internal/offset/model.go:53]: It's being sanitized during publish. But as a precaustion maybe we sanitize here as well?

File: internal/subscriber/ordering.go [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+	return s.subscription
+}
+
+// CanConsumeMore looks at sum of all consumed messages in all the active topic partitions and checks threshold
+func (s *OrderedImplementation) CanConsumeMore() bool {
+	totalConsumedMsgsForTopic := 0
+	for _, cm := range s.consumedMessageStats {
+		totalConsumedMsgsForTopic += len(cm.consumedMessages)
+	}
+	return totalConsumedMsgsForTopic <= int(s.maxOutstandingMessages)
```
[the remark this answers is not in the document]
Comment (razorao) [reply internal/subscriber/ordering.go:52]: Yes. Not being done right now. But this should be taken from the pull request.

File: internal/subscription/validation.go [hunk 4]
```diff
@@ -142,6 +142,8 @@ func GetValidatedModelForCreate(ctx context.Context, req *metrov1.Subscription)
 		return nil, merror.Newf(merror.InvalidArgument, err.Error())
 	}
 
+	m.EnableMessageOrdering = req.EnableMessageOrdering
```
[the remark this answers is not in the document]
Comment (razorao) [reply internal/subscription/validation.go:145]: You mean an ordered subscription turning into unordered? That's not allowed. We don't allow any edits.

**Pull request merged**